rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala

```diff
@@ -0,0 +1,173 @@
+package com.stripe.rainier.core
+
+import com.stripe.rainier.sampler._
+import com.stripe.rainier.compute._
+
+//implements Simulation-Based Calibration from https://arxiv.org/abs/1804.06788
+final case class SBC[T](priorGenerators: Seq[Generator[Double]],
+                        priorParams: Seq[Real],
+                        posterior: RandomVariable[(Distribution[T], Real)]) {
+
+  import SBC._
+
+  require(priorParams.forall(_.variables.size == 1))
+  val priorGenerator = Generator.traverse(priorGenerators)
+  val emptyEvaluator = new Evaluator(Map.empty)
+
+  def prepare(sampler: Sampler, warmupIterations: Int, syntheticSamples: Int)(
+      implicit rng: RNG): Run = {
+    val initialThin = 2
+    val (_, maxRHat, minEffectiveSampleSize) =
+      sample(sampler, warmupIterations, syntheticSamples, initialThin)
+    val thin =
+      if (minEffectiveSampleSize < Samples)
+        Math
+          .ceil(initialThin.toDouble * Samples / minEffectiveSampleSize)
+          .toInt
+      else
+        initialThin
+    Run(sampler, warmupIterations, syntheticSamples, thin, maxRHat)
+  }
+
+  private def sample(sampler: Sampler,
+                     warmupIterations: Int,
+                     syntheticSamples: Int,
+                     thin: Int)(implicit rng: RNG): (Int, Double, Double) = {
+    val trueValues = priorGenerator.get(rng, emptyEvaluator)
+    implicit val trueEval = new Evaluator(priorParams.zip(trueValues).toMap)
+    val trueOutput = posterior.map(_._2).get
+
+    val syntheticValues =
+      posterior.map(_._1.generator.repeat(syntheticSamples)).get
+    val model = posterior.flatMap {
+      case (d, r) =>
+        d.fit(syntheticValues).map { _ =>
+          r
+        }
+    }
+    val (samples, diag) = model.sampleWithDiagnostics(sampler,
+                                                      Chains,
+                                                      warmupIterations,
+                                                      (Samples / Chains) * thin,
+                                                      thin)
+
+    val maxRHat = diag.map(_.rHat).max
+    val minEffectiveSampleSize = diag.map(_.effectiveSampleSize).min
+    val rawRank = samples.tail.count { n =>
```

Comment:
Why the `tail`?

Reply:
We have some `L` which is the number of possible values we want `rawRank` to be able to take on. (As it happens we hard code that to `128`). Because the true value can be before the first sample, or after the last sample (or anywhere in between), this means we only want to compare it against `L-1` different samples. But we're producing our samples from 4 different chains, which means we'd need to sample `(L-1)/4` from each chain. Since we can't sample fractional samples, and since we know that `L` in (and always should be) evenly divisible by 4, we just sample `L/4` (= 32) from each chain and then throw one out to end up with `L-1`.

Reply:
Also we throw out the first one in particular because it's the least likely one to be converged.

Reply:
Ah, thanks. That fractional sample subtlety escaped me.